lib/ASTGen/Sources/ASTGen/Bridge.swift

```diff
@@ -318,3 +318,20 @@ extension BridgedArrayRef {
     return c(buffer)
   }
 }
+
+/// Utility to pass Swift closure to C/C++ bridging API.
+///
+/// C/C++ API can call the closure via `BridgedSwiftClosure::operator()` which
+/// calls `bridgedSwiftClosureCall_1(_:_:)` function below.
+func withBridgedSwiftClosure(closure: (UnsafeRawPointer?) -> Void, call: (BridgedSwiftClosure) -> Void) {
+  withoutActuallyEscaping(closure) { escapingClosure in
+    withUnsafePointer(to: escapingClosure) { ptr in
+      call(BridgedSwiftClosure(closure: ptr))
+    }
+  }
+}
+
+@_cdecl("swift_ASTGen_bridgedSwiftClosureCall_1")
+func bridgedSwiftClosureCall_1(_ bridged: BridgedSwiftClosure, _ arg: UnsafeRawPointer?) {
+  bridged.closure.assumingMemoryBound(to: ((UnsafeRawPointer?) -> Void).self).pointee(arg)
+}
```

Comment:
Can't say I love this, I guess the alternative would be having the bridging API take a C function pointer along with an opaque `void *` context which the caller would then load as the right type. Not sure that would be much of an improvement though :(

Reply:
I don't love either, this was the best I came up with 😓